src/utils/records.ts

```diff
@@ -58,6 +56,51 @@ export class Records {
     return plaintextStream;
   }
 
+  /**
+   * Constructs full key derivation path using the specified scheme.
+   */
+  public static constructKeyDerivationPath(
+    _keyDerivationScheme: KeyDerivationScheme,
+    recordId: string,
+    contextId: string | undefined,
+    descriptor: RecordsWriteDescriptor
+  ): string[] {
+
+    // NOTE: right now only `protocols` derivation scheme is supported so we will assume that's the scheme without additional switch/if statements
+    const fullDerivationPath = Records.constructKeyDerivationPathUsingProtocolsScheme(recordId, contextId, descriptor);
+    return fullDerivationPath;
+  }
+
+  /**
+   * Constructs the full key derivation path using `protocols` scheme.
+   */
+  private static constructKeyDerivationPathUsingProtocolsScheme(
+    recordId: string,
+    contextId: string | undefined,
+    descriptor: RecordsWriteDescriptor
+  ): string[] {
+    // ensure `protocol` is defined
+    // NOTE: no need to check `protocolPath` and `contextId` because earlier code ensures that if `protocol` is defined, those are defined also
+    if (descriptor.protocol === undefined) {
+      throw new DwnError(
+        DwnErrorCode.RecordsProtocolsDerivationSchemeMissingProtocol,
+        'Unable to construct key derivation path using `protocols` scheme because `protocol` is missing.'
+      );
+    }
+
+    const protocolPathSegments = descriptor.protocolPath!.split('/');
+    const fullDerivationPath = [
+      KeyDerivationScheme.Protocols,
+      descriptor.protocol,
+      contextId!,
```

Comment:
Nit: why bother making allowing `contextId` to be undefined if we're just gonna `!`? I didn't see at a glance the spot between here and the top of `RecordsWrite#handler` that we ensure `contextId` is in fact defined.

Reply:
Ah wait I see it. In `RecordsWrite#handler`, we do 
```typescript
if (descriptor.protocol !== undefined) {
  contextId = await RecordsWrite.getEntryId(author, descriptor);
}
```
Do you mind making `contextId: string` inside `constructKeyDerivationPathUsingProtocolsScheme` and doing the `context!` in `constructKeyDerivationPath` on line 70 instead?

Reply:
@diehuxx, but `context` is only guaranteed defined if `protocol` is defined, and in `constructKeyDerivationPath` we don't know if `protocol` is defined or not, unless we move the `if (descriptor.protocol === undefined)` into `constructKeyDerivationPath` also, but then I intentionally wanted to hide scheme specific checks within scheme specific construction methods.

Reply:
Ah that's fair. Agreed we def should keep the `contextId!` assertion after we check whether `protocol` is defined.